### PR TITLE
Add PEP 706 filters to tarfile

### LINKED ISSUE #10315: Daily tests failed on Wed Jun 14 2023
Runs listed here: https://github.com/python/typeshed/actions/workflows/daily.yml

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Fixes #10315

### CODE REVIEW

**Comment by srittau** [comment]
It seems that GitHub has already updated some of its runners to newer Python versions, but not all, causing intermittent Actions failures.

**Comment by JelleZijlstra** [comment]
Yes, let's leave this open until all the runners are updated.

**Comment by AlexWaygood** [comment]
> Yes, let's leave this open until all the runners are updated.

That might be a while -- in the meantime, we could maybe use a hack like this to get stubtest to quiet down about this in CI?

https://github.com/python/typeshed/blob/6e90fce4b88e9ea4b4f79070f83bbbae7bcf641d/tests/stubtest_allowlists/py3_common.txt#L612-L615

**Comment by JelleZijlstra** [comment]
CI is green now.

File: stdlib/tarfile.pyi [hunk 1]
```diff
@@ -324,6 +360,29 @@ class StreamError(TarError): ...
 class ExtractError(TarError): ...
 class HeaderError(TarError): ...
 
+if sys.version_info >= (3, 8):
+    class FilterError(TarError):
+        tarinfo: TarInfo
+
+    class AbsolutePathError(FilterError):
+        def __init__(self, tarinfo: TarInfo) -> None: ...
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply stdlib/tarfile.pyi:368]: Hmmmmmm, looks like the idea is that `FilterError` is meant to be an abstract base class that is never actually instantiated from within `tarfile` nor by users. So "instances of `FilterError`" will in practice actually be "instances of subclasses of `FilterError`", which _are_ guaranteed to have the attribute. So I guess this is okay (though a comment probably wouldn't go amiss :)

File: stdlib/tarfile.pyi [hunk 2]
```diff
@@ -324,6 +360,29 @@ class StreamError(TarError): ...
 class ExtractError(TarError): ...
 class HeaderError(TarError): ...
 
+if sys.version_info >= (3, 8):
+    class FilterError(TarError):
+        tarinfo: TarInfo
```
Comment (JelleZijlstra) [inline stdlib/tarfile.pyi:365]:
```suggestion
    class FilterError(TarError):
        # This attribute is only set directly on the subclasses, but the documentation guarantees
        # that it is always present on FilterError.
        tarinfo: TarInfo
```

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**